flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/plan/nodes/exec/testutils/JoinTestPrograms.java

```diff
@@ -0,0 +1,272 @@
+                                                Row.of("Hello world", "Hallo Welt"),
+                                                Row.of("Hi", "Hallo Welt wie gehts"))
+                                        .consumedAfterRestore(
+                                                Row.of("Hello there", "Hallo Welt wie"))
+                                        .build())
+                        .runSql("insert into MySink " + "SELECT a3, b4 FROM A, B WHERE a2 = b2")
+                        .build();
+
+        INNER_JOIN =
+                TableTestProgram.of("inner join", "test inner join")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a1 int", "b1 int")
+                                        .consumedBeforeRestore(
+                                                Row.of(1, 1),
+                                                Row.of(2, 2),
+                                                Row.of(3, 3),
+                                                Row.of(2, 2))
+                                        .consumedAfterRestore(Row.of(2, 2))
+                                        .build())
+                        .runSql("insert into MySink " + "SELECT a1, b1 FROM A JOIN B ON a1 = b1")
+                        .build();
+
+        JOIN_WITH_FILTER =
+                TableTestProgram.of("join-with-filter", "test join with filter")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a3 varchar", "b4 varchar")
+                                        .consumedBeforeRestore(
+                                                Row.of("Hi", "Hallo"),
+                                                Row.of("Hi", "Hallo Welt wie gehts"))
+                                        .consumedAfterRestore(new Row[] {})
+                                        .build())
+                        .runSql(
+                                "insert into MySink "
+                                        + "SELECT a3, b4 FROM A, B where a2 = b2 and a2 < 2")
+                        .build();
+
+        INNER_JOIN_WITH_DUPLICATE_KEY =
+                TableTestProgram.of(
+                                "inner-join-with-duplicate-key", "inner join with duplicate key")
+                        .setupTableSource(SOURCE_A)
+                        .setupTableSource(SOURCE_B)
+                        .setupTableSink(
+                                SinkTestStep.newBuilder("MySink")
+                                        .addSchema("a1 int", "b1 int")
+                                        .consumedBeforeRestore(Row.of(2, 2))
```

Comment:
can we use columns that would tell us which rows actually got joined?

Reply:
I think the switch to employees and departments will help with this!